app/src/main/java/com/oracle/javafx/scenebuilder/app/welcomedialog/WelcomeDialogWindowController.java

```diff
@@ -141,17 +171,97 @@ public static WelcomeDialogWindowController getInstance() {
         return instance;
     }
 
+    private void fireSelectTemplate(Template template) {
+        if (sceneBuilderApp.startupTasksFinishedProperty().get()) {
+            sceneBuilderApp.performNewTemplate(template);
+            getStage().hide();
+        } else {
+            showMasker(() -> {
+                sceneBuilderApp.performNewTemplate(template);
+                getStage().hide();
+            });
+        }
+    }
+
     private void fireOpenRecentProject(ActionEvent event, String projectPath) {
-        sceneBuilderApp.handleOpenFilesAction(Arrays.asList(projectPath));
-        getStage().hide();
+        handleOpen(List.of(projectPath));
     }
 
     @FXML
     private void openDocument() {
-        // Right now there is only one window open by default
-        DocumentWindowController documentWC = sceneBuilderApp.getDocumentWindowControllers().get(0);
-        sceneBuilderApp.performControlAction(SceneBuilderApp.ApplicationControlAction.OPEN_FILE, documentWC);
-        getStage().hide();
+        FileChooser fileChooser = new FileChooser();
+        fileChooser.getExtensionFilters().add(
+                new FileChooser.ExtensionFilter(I18N.getString("file.filter.label.fxml"), "*.fxml")
+        );
+        fileChooser.setInitialDirectory(EditorController.getNextInitialDirectory());
+
+        List<File> fxmlFiles = fileChooser.showOpenMultipleDialog(getStage());
+
+        // no file was selected, so nothing to do
+        if (fxmlFiles == null)
+            return;
+
+        List<String> paths = fxmlFiles
+                .stream()
+                .map(File::toString)
+                .collect(Collectors.toList());
+
+        handleOpen(paths);
+    }
+
+    private void handleOpen(List<String> paths) {
+        if (sceneBuilderApp.startupTasksFinishedProperty().get()) {
+            sceneBuilderApp.handleOpenFilesAction(paths);
+            getStage().hide();
+        } else {
+            showMasker(() -> {
+                sceneBuilderApp.handleOpenFilesAction(paths);
+                getStage().hide();
+            });
+        }
+    }
+
+    private void showMasker(Runnable onEndAction) {
+        // swap roots to allow adding masker if running for the first time
+        if (newRoot == null) {
+            initMasker();
+        }
+
+        rootPane.setDisable(true);
+        newRoot.getChildren().add(masker);
+
+        sceneBuilderApp.startupTasksFinishedProperty().addListener((o, old, isFinished) -> {
+            if (isFinished) {
+                Platform.runLater(() -> {
+                    onEndAction.run();
+
+                    // restore root in case welcome dialog is opened again
+                    rootPane.setDisable(false);
+                    newRoot.getChildren().remove(masker);
+                });
+            }
+        });
+    }
+
+    private void initMasker() {
+        newRoot = new StackPane();

```

Comment:
I guess this is okay for now, but Scene Builder tends to build itself from FXML files. It might be easier to have the masker already as an existing invisible node in `WelcomeWindow.fxml`, that can get activated as needed.

Reply:
Given the masker does not need to be initialized for all cases, I think the lazy init is fine here.